File: Sources/ExperimentKit/ExecutionPlan.swift
import Foundation

/// What a study will ACTUALLY do when it runs — one resolver, used by the run
/// loops, the routing rules and the UI (E1).
///
/// The rule "which instruments imply generation" already existed, but only
/// inside the run loop (`wantsSampled` / `wants_sampled`). Everything else
/// re-derived its own answer or guessed. The visible consequence: a
/// logprob-only study was **pinned to the server by a temperature its
/// instrument ignores**. `LogprobInstrument` scores options through a stepped
/// KV cache and never samples, so temperature decides nothing for it — yet a
/// declared `temperature > 0` made the study server-only, because the routing
/// rule looked at the number without asking whether anything read it.
///
/// So routing depends on the PLAN, not on a field that may be inert. And
/// where a declared value is inert, that is said out loud rather than
/// silently ignored — a temperature that decides nothing is a study-design
/// mistake worth surfacing, just not a reason to move the study to another
/// substrate.
///
/// Cross-engine twin: `Server/steerlab_server/experiment/execution_plan.py`.
public enum ExecutionPlan {

    /// Instruments that require sampled generation.
    ///
    /// `repeReaderScore` reads a reader over the model's OUTPUT, so it needs
    /// text even though it is not itself a sampled-text metric.
    public static let generationImplyingInstruments: Set<String> = [
        "sampledText", "repeReaderScore",
    ]

    /// Instruments that read option logprobs directly, with no generation.
    public static let directScoringInstruments: Set<String> = [
        "answerTokenLogprob", "choiceProbability", "ordinalScale",
    ]

    public struct Plan: Sendable, Equatable {
        /// The study samples text. Temperature and seeds are operative only
        /// when this is true.
        public var generatesSampledText: Bool
        /// The study scores declared options deterministically.
        public var scoresDirectly: Bool
        /// The resolved instrument list (empty input resolves to sampled
        /// text, the engine default).
        public var instruments: [String]

        /// Sampling settings are operative for this plan.
        public var samplingIsOperative: Bool { generatesSampledText }

        public var summary: String {
            switch (generatesSampledText, scoresDirectly) {
            case (true, true):
                "generates sampled text AND scores options deterministically"
            case (true, false):
                "generates sampled text"
            case (false, true):
                "scores options deterministically — no sampling"
            case (false, false):
                "nothing to run"
            }
        }
    }

    /// Resolve what the declared instruments mean.
    ///
    /// Absent or empty resolves to sampled text — the engine default both run
    /// loops already apply (`instruments.isEmpty || contains("sampledText")`).
    public static func resolve(instruments: [String]?) -> Plan {
        let declared = instruments ?? []
        guard !declared.isEmpty else {
            return Plan(
                generatesSampledText: true, scoresDirectly: false,
                instruments: ["sampledText"])
        }
        let set = Set(declared)
        return Plan(
            generatesSampledText: !set.isDisjoint(with: generationImplyingInstruments),
            scoresDirectly: !set.isDisjoint(with: directScoringInstruments),
            instruments: declared)
    }

    /// A declared sampling setting that this plan will not read.
    ///
    /// Advisory, never a refusal: declaring a temperature on a
    /// deterministic-only study is a design mistake, but the RUN is
    /// well-defined and its result is unaffected.
    public static func inertSamplingAdvisory(
        instruments: [String]?, temperature: Double?, samplesPerItem: Int?
    ) -> String? {
        let plan = resolve(instruments: instruments)
        guard !plan.samplingIsOperative else { return nil }
        var inert: [String] = []
        if (temperature ?? 0) > 0 {
            inert.append(
                "temperature "
                    + (temperature ?? 0).formatted(
                        .number.precision(.fractionLength(0 ... 2))))
        }
        if (samplesPerItem ?? 1) > 1 {
            inert.append("samplesPerItem \(samplesPerItem ?? 1)")
        }
        guard !inert.isEmpty else { return nil }
        return "this study \(plan.summary), so \(inert.joined(separator: " and ")) "
            + "\(inert.count == 1 ? "is" : "are") inert — the deterministic "
            + "instruments never sample, so the declared value changes "
            + "nothing. Remove it, or add a generating instrument if sampled "
            + "output was intended."
    }
}

File: Server/steerlab_server/experiment/execution_plan.py
"""What a study will ACTUALLY do when it runs — one resolver (E1).

The rule "which instruments imply generation" already existed, but only inside
the run loop (``wants_sampled`` / Swift ``wantsSampled``). Everything else
re-derived its own answer or guessed. The visible consequence: a logprob-only
study was **pinned to the server by a temperature its instrument ignores**.
The logprob instrument scores options through a stepped KV cache and never
samples, so temperature decides nothing for it — yet a declared
``temperature > 0`` made the study server-only, because the routing rule read
the number without asking whether anything downstream reads it.

So routing depends on the PLAN, not on a field that may be inert. Where a
declared value IS inert, that is said out loud rather than silently ignored: a
temperature that decides nothing is a study-design mistake worth surfacing,
just not a reason to move the study to another substrate.

Cross-engine twin: ``Sources/ExperimentKit/ExecutionPlan.swift``.
"""

from __future__ import annotations

from dataclasses import dataclass, field

#: Instruments that require sampled generation. ``repeReaderScore`` reads a
#: reader over the model's OUTPUT, so it needs text even though it is not
#: itself a sampled-text metric.
GENERATION_IMPLYING_INSTRUMENTS = frozenset({"sampledText", "repeReaderScore"})

#: Instruments that read option logprobs directly, with no generation.
DIRECT_SCORING_INSTRUMENTS = frozenset(
    {"answerTokenLogprob", "choiceProbability", "ordinalScale"})


@dataclass(frozen=True)
class Plan:
    generates_sampled_text: bool
    scores_directly: bool
    instruments: list = field(default_factory=list)

    @property
    def sampling_is_operative(self) -> bool:
        """Temperature and seeds are operative only when something samples."""
        return self.generates_sampled_text

    @property
    def summary(self) -> str:
        if self.generates_sampled_text and self.scores_directly:
            return "generates sampled text AND scores options deterministically"
        if self.generates_sampled_text:
            return "generates sampled text"
        if self.scores_directly:
            return "scores options deterministically — no sampling"
        return "nothing to run"


def resolve(instruments) -> Plan:
    """Absent or empty resolves to sampled text — the engine default both run
    loops already apply."""
    declared = list(instruments or [])
    if not declared:
        return Plan(True, False, ["sampledText"])
    as_set = set(declared)
    return Plan(
        bool(as_set & GENERATION_IMPLYING_INSTRUMENTS),
        bool(as_set & DIRECT_SCORING_INSTRUMENTS),
        declared)


def inert_sampling_advisory(instruments, temperature, samples_per_item
                            ) -> str | None:
    """A declared sampling setting this plan will not read.

    Advisory, never a refusal: declaring a temperature on a
    deterministic-only study is a design mistake, but the RUN is well-defined
    and its result unaffected."""
    plan = resolve(instruments)
    if plan.sampling_is_operative:
        return None
    inert = []
    if (temperature or 0) > 0:
        inert.append(f"temperature {temperature:g}")
    if (samples_per_item or 1) > 1:
        inert.append(f"samplesPerItem {samples_per_item}")
    if not inert:
        return None
    verb = "is" if len(inert) == 1 else "are"
    return (f"this study {plan.summary}, so {' and '.join(inert)} {verb} inert "
            "— the deterministic instruments never sample, so the declared "
            "value changes nothing. Remove it, or add a generating instrument "
            "if sampled output was intended.")
